cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,492 @@
+                          num_rows,
+                          mask_filter{decode_kernel_mask::FIXED_WIDTH_DICT},
+                          page_processing_stage::DECODE)) {
+    return;
+  }
+
+  // the level stream decoders
+  // rolling_buf_size = 256
+  __shared__ rle_run<level_t> def_runs[rle_run_buffer_size];
+  rle_stream<level_t, decode_block_size, rolling_buf_size> def_decoder{def_runs};
+
+  // should the size be 1/2 (128?)
+  __shared__ rle_run<uint32_t> dict_runs[rle_run_buffer_size];  // should be array of 6
+  rle_stream<uint32_t, decode_block_size, rolling_buf_size> dict_stream{dict_runs};
+
+  bool const nullable = s->col.max_level[level_type::DEFINITION] > 0;
+
+  // if we have no work to do (eg, in a skip_rows/num_rows case) in this page.
+  //
+  // corner case: in the case of lists, we can have pages that contain "0" rows if the current row
+  // starts before this page and ends after this page:
+  //       P0        P1        P2
+  //  |---------|---------|----------|
+  //        ^------------------^
+  //      row start           row end
+  // P1 will contain 0 rows
+  //
+  if (s->num_rows == 0) { return; }
+
+  // initialize the stream decoders (requires values computed in setupLocalPageInfo)
+  level_t* def = reinterpret_cast<level_t*>(pp->lvl_decode_buf[level_type::DEFINITION]);
+  if (nullable) {
+    // col.level_bits - 1
+    def_decoder.init(s->col.level_bits[level_type::DEFINITION],
+                     s->abs_lvl_start[level_type::DEFINITION],
+                     s->abs_lvl_end[level_type::DEFINITION],
+                     def,  // 2048 sized
+                     s->page.num_input_values);
+  }
+
+  dict_stream.init(
+    s->dict_bits, s->data_start, s->data_end, sb->dict_idx, s->page.num_input_values);
+  __syncthreads();
+
+  // the core loop. decode batches of level stream data using rle_stream objects
+  // and pass the results to gpuUpdatePageSizes
+  int processed = 0;
+  int valid     = 0;
+  while (processed < s->page.num_input_values) {
+    int next_valid;
+
+    // only need to process definition levels if this is a nullable column
+    if (nullable) {
+      //-1 don't cap it
+      processed += def_decoder.decode_next(t);
+      __syncthreads();
+
+      // count of valid items in this batch
+      next_valid =
+        gpuUpdateValidityOffsetsAndRowIndicesFlat<true, level_t>(processed, s, sb, def, t);
+    }
+    // if we wanted to split off the skip_rows/num_rows case into a separate kernel, we could skip
+    // this function call entirely since all it will ever generate is a mapping of (i -> i) for
+    // nz_idx.  gpuDecodeValues would be the only work that happens.
+    else {
+      processed += min(rolling_buf_size, s->page.num_input_values - processed);
+      next_valid =
+        gpuUpdateValidityOffsetsAndRowIndicesFlat<false, level_t>(processed, s, sb, nullptr, t);
+    }
+    __syncthreads();
+
+    dict_stream.decode_next(t, (next_valid - valid));
```

Comment:
A comment on the significance of `next_valid - valid` would be useful here.  Actually, as a general comment for this kernel and the other one, the significance of value count vs valid value count is probably worth a nice block comment.

Reply:
I added a comment for this @nvdbaranec. Let me know if that looks ok

Reply:
I'm not seeing it.

Reply:
I added this https://github.com/rapidsai/cudf/pull/15159/files#diff-7f915e59aaa858b005b7d4ba9166e81015d1452cc1f655cd43ffc818fbc82375R461 for `next_valid_count - valid_count`, and this https://github.com/rapidsai/cudf/pull/15159/files#diff-7f915e59aaa858b005b7d4ba9166e81015d1452cc1f655cd43ffc818fbc82375R421-R428 for an explanation of `valid_count` vs `processed_count`